## Review — PR #27 (round 4) — ⚠️ blocking regression in CI

Re-opening: CI is **red** on `c6a2af8` (`mergeStateStatus: UNSTABLE`). Builds (all 6 platforms) and Lint pass, but the **Test** job fails:

```
=== RUN   TestRunCLIIdleKill
    cli_provider_test.go:114: idle kill took 5.0023237s, expected to fire near the 150ms window
--- FAIL: TestRunCLIIdleKill (5.00s)
FAIL	github.com/srvsngh99/mini-krill/internal/llm	10.541s
```

It passes locally on macOS (`0.16s`) and fails on the Linux CI runner — and this is **not flaky-timing**: the test is correctly catching a real defect in the idle-kill path.

### Substantive issue (new — round 4, item 3)

**`runCLIWithIdle` does not reliably reclaim a stuck process within the idle window when the subprocess has children** — `cli_provider.go` idle-monitor + `<-done` gate.

Diagnosis: on idle fire the monitor does `idled.Store(true); cancel()`. `cancel()` cancels the child context, and `exec.CommandContext` sends SIGKILL to **`cmd.Process` only** — the direct child. `sh -c "sleep 5"` on the Linux runner does *not* exec-replace into `sleep`; it forks a `sleep` grandchild. SIGKILL reaps `sh` but the orphaned `sleep` survives and **inherits the write ends of the stdout/stderr pipes**. The reader goroutines' `pipe.Read` therefore does not EOF until *every* writer FD closes — i.e. not until `sleep` exits 5s later. Since round 2 deliberately gates `cmd.Wait()` behind `<-done` (both readers returned), the whole function blocks for the full natural runtime of the orphaned grandchild. The "10-minute idle kill" silently degrades to "however long the stuck child tree lives" — which defeats the feature for exactly the real Codex/Claude CLIs that spawn helper subprocesses (the PR's whole motivation). macOS happens to exec-replace `sh`→`sleep` (one PID), which is why local passed and masked it.

**Fix direction (either, not both required):**

1. *Process-group kill (most direct — keeps the round-2 `<-done` gate working as designed):* start the child in its own process group — `cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}` — and on cancel signal the whole group (`syscall.Kill(-pgid, SIGKILL)`), so grandchildren die too, the pipe writers all close, and `<-done` unblocks promptly. This is Unix-only; gate it behind a `_unix.go` / `_windows.go` build-tag split (the tests already `skipIfNoShell` on Windows, and the target CLIs are POSIX).
2. *Idiomatic os/exec route:* set `cmd.Cancel` + `cmd.WaitDelay` (Go 1.20+) and call `cmd.Wait()` directly instead of gating on `<-done`. `WaitDelay` bounds how long `Wait` blocks on still-open inherited pipes after the process is signalled, then force-closes them — which is precisely the "orphan holds the FD" case. Note this only helps if the `<-done` gate is removed, since today the code never reaches `cmd.Wait()` while a grandchild keeps a reader alive; `WaitDelay` + direct `Wait()` also re-satisfies the StdoutPipe ordering contract that motivated round 2's gate, so it's a clean structural replacement rather than a regression.

I'd lean to option 1: smallest change, preserves the round-2 design, and `Setpgid`+group-kill is the standard way to kill a subprocess *tree*.

### Not regressed (round 1–3 items still good)

- Single-owner timer (round-1 item 1) — still correct; `TestRunCLISlowStreamSurvives` and `-race` still pass.
- `<-done` before `cmd.Wait()` (round-1 item 2) — the *ordering* is still right; the new finding is that `<-done` itself can hang on an orphaned grandchild, which is a different defect, not a regression of the round-2 fix.
- `ctx.Err()` surfacing + `idleTimeout` doc (round-1 nits) — intact; `TestRunCLIParentCancel` passes.

### Verdict

**Re-requesting changes for round-4 item 3** (idle-kill does not reclaim a child-spawning subprocess within the window; CI Test red). Once the process-group / `WaitDelay` fix lands and `TestRunCLIIdleKill` is green in CI, this is back to LGTM — rounds 1–3 remain validated. Holding the merge until CI is green.

--- a/internal/llm/cli_provider.go
+++ b/internal/llm/cli_provider.go
@@ -151,6 +151,19 @@ func runCLIWithIdle(ctx context.Context, name string, args []string, stdin strin
 		cmd.Stdin = strings.NewReader(stdin)
 	}
 
+	// Run the child in its own process group and, on context cancel (idle
+	// kill or parent cancel), SIGKILL the whole group rather than just the
+	// direct child. exec.CommandContext's default Cancel kills cmd.Process
+	// only; a CLI that forks helper grandchildren would leave them holding
+	// the pipe write ends, so the reader goroutines never EOF and <-done
+	// (and thus the idle-kill) hangs until the orphan exits on its own.
+	setProcessGroup(cmd)
+	cmd.Cancel = func() error { return killProcessTree(cmd) }
+	// Belt-and-suspenders bound on cmd.Wait(): if a stray inherited FD is
+	// still open after the group kill, force the pipes closed rather than
+	// blocking Wait indefinitely.
+	cmd.WaitDelay = 10 * time.Second
+
 	// Suppress macOS dyld MallocStackLogging warnings at the source. The flag
 	// is leaked into every Go-spawned subprocess on dev machines and pollutes
 	// dive.log with thousands of "can't turn off malloc stack logging" lines.

--- a/internal/llm/cli_provider_test.go
+++ b/internal/llm/cli_provider_test.go
@@ -115,6 +115,30 @@ func TestRunCLIIdleKill(t *testing.T) {
 	}
 }
 
+// A silent subprocess that forks a grandchild must still be reclaimed
+// within the idle window. The compound command stops `sh` from
+// exec-replacing into `sleep`, so `sleep` is a forked grandchild that
+// inherits the pipe write ends. Killing only the direct child (the
+// pre-fix behaviour) would orphan `sleep`, keep the pipes open, and hang
+// <-done for the full 5s — the round-4 CI regression. The process-group
+// kill must take the whole tree down promptly.
+func TestRunCLIIdleKillReclaimsChildTree(t *testing.T) {
+	skipIfNoShell(t)
+	start := time.Now()
+	_, err := runCLIWithIdle(context.Background(), "sh",
+		[]string{"-c", "sleep 5; :"}, "", 150*time.Millisecond)
+	if err == nil {
+		t.Fatal("runCLIWithIdle: expected idle-kill error, got nil")
+	}
+	if !strings.Contains(err.Error(), "idle for") {
+		t.Fatalf("runCLIWithIdle error = %v, want idle-kill message", err)
+	}
+	// Must fire near the 150ms window, not the grandchild's 5s lifetime.
+	if elapsed := time.Since(start); elapsed > 2*time.Second {
+		t.Fatalf("idle kill took %v — orphaned grandchild not reclaimed", elapsed)
+	}
+}
+
 // A subprocess that keeps streaming output, each chunk within the idle
 // window, must run to completion across many idle windows without being
 // killed — the exact multi-step-plan scenario the PR targets.

--- a/internal/llm/cli_provider_unix.go
+++ b/internal/llm/cli_provider_unix.go
@@ -1,0 +1,38 @@
+//go:build !windows
+
+package llm
+
+import (
+	"os/exec"
+	"syscall"
+)
+
+// setProcessGroup makes the child the leader of a new process group, so the
+// whole subprocess tree can be signalled at once. The target Codex/Claude
+// CLIs (and `sh -c` in the tests) fork helper grandchildren; without this,
+// exec.CommandContext SIGKILLs only the direct child, the orphaned
+// grandchildren keep the stdout/stderr pipe write ends open, the reader
+// goroutines never EOF, and the idle-kill path hangs on <-done until the
+// grandchild exits on its own — the round-4 CI regression.
+func setProcessGroup(cmd *exec.Cmd) {
+	if cmd.SysProcAttr == nil {
+		cmd.SysProcAttr = &syscall.SysProcAttr{}
+	}
+	cmd.SysProcAttr.Setpgid = true
+}
+
+// killProcessTree SIGKILLs the entire process group led by cmd's child. With
+// Setpgid the child's PGID equals its PID, so -PID addresses the whole group
+// (children and grandchildren included). Safe to call repeatedly and after
+// the process has already exited: a vanished group yields ESRCH, which is
+// ignored. Falls back to killing just the direct child if the group signal
+// fails for any other reason.
+func killProcessTree(cmd *exec.Cmd) error {
+	if cmd.Process == nil {
+		return nil
+	}
+	if err := syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL); err != nil && err != syscall.ESRCH {
+		return cmd.Process.Kill()
+	}
+	return nil
+}

--- a/internal/llm/cli_provider_windows.go
+++ b/internal/llm/cli_provider_windows.go
@@ -1,0 +1,17 @@
+//go:build windows
+
+package llm
+
+import "os/exec"
+
+// setProcessGroup is a no-op on Windows: there is no POSIX process group, the
+// target CLIs are POSIX, and the shell-based runCLI tests skip on Windows.
+func setProcessGroup(cmd *exec.Cmd) {}
+
+// killProcessTree falls back to killing the direct child on Windows.
+func killProcessTree(cmd *exec.Cmd) error {
+	if cmd.Process == nil {
+		return nil
+	}
+	return cmd.Process.Kill()
+}

